components/google_sheets/sources/common/http-based/base.mjs

```diff
@@ -0,0 +1,141 @@
+// This source has two interfaces:
+//
+// 1) The HTTP requests tied to changes in a Google Drive
+// 2) A timer that runs on regular intervals, renewing the notification channel
+//    as needed
+//
+// This is the base module for other Google Drive subscription sources, which
+// should override the unimplemented methods, to create a webhook and process
+// incoming events.
+
+import { WEBHOOK_SUBSCRIPTION_RENEWAL_SECONDS } from "../../../../google_drive/constants.mjs";
+import googleSheets from "../../../google_sheets.app.mjs";
+import { MY_DRIVE_VALUE } from "../../../../google_drive/constants.mjs";
+
+/**
+ * The number of events that will be automatically sent whenever the event
+ * source is setup and deployed for the first time.
+ *
+ * Note that the event source could send less initial events than this if the
+ * associated worksheets do not contain enough data.
+ */
+const INITIAL_EVENT_COUNT = 10;
+
+/* eslint-disable pipedream/required-properties-key, pipedream/required-properties-name,
+  pipedream/required-properties-version, pipedream/required-properties-description,
+  pipedream/required-properties-type */
+export default {
+  props: {
+    googleSheets,
+    db: "$.service.db",
+    http: "$.interface.http",
+    timer: {
+      label: "Push notification renewal schedule",
+      description:
+        "The Google Drive API requires occasionally renewal of push notification subscriptions. **This runs in the background, so you should not need to modify this schedule**.",
+      type: "$.interface.timer",
+      static: {
+        intervalSeconds: WEBHOOK_SUBSCRIPTION_RENEWAL_SECONDS,
+      },
+    },
+  },
+  hooks: {
+    async deploy() {
+      await this.takeSheetSnapshot(INITIAL_EVENT_COUNT);
+
+      const sheetId = this.getSheetId();
+      const spreadsheet = await this.googleSheets.getSpreadsheet(sheetId);
+      await this.processSpreadsheet(spreadsheet);
+    },
+    async activate() {
+      throw new Error("activate is not implemented");
+    },
```

Comment:
Since `activate()` is so similar in drive.mjs and sheet.mjs, I think we could put it in base.mjs and use methods `activateHook` in both drive.mjs and sheet.mjs for the call to either `this.googleSheets.activateHook()` or `this.googleSheets.activateFileHook()`.
```suggestion
    async activate() {
      const channelID = this._getChannelID() || uuid();

      const {
        startPageToken,
        expiration,
        resourceId,
      } = await this.activateHook(channelID);

      // Save metadata on the subscription so we can stop / renew later
      // Subscriptions are tied to Google's resourceID, "an opaque value that
      // identifies the watched resource". This value is included in request headers
      this._setSubscription({
        resourceId,
        expiration,
      });
      this._setChannelID(channelID);
      this._setPageToken(startPageToken);

      await this.takeSheetSnapshot();
    },
```